### PR TITLE
Feat/refactor library

### PR DESCRIPTION
**Pull request by christianscott**
Wee bit of a refactor. No need to merge this, just though I'd tidy things up a bit. The last commit is probably the juiciest, feel free to ping me if anything is confusing 😄

### CODE REVIEW

File: src/RedditPosts.js [hunk 1]
```diff
@@ -1,153 +1,126 @@
...
+
+    this.componentChose = this.componentChose.bind(this);
+    this.mediaComprehension = this.getMediaFromPost.bind(this);
+    this.handlePosts = this.handlePosts.bind(this);
   }
 
-  this.componentChose = this.componentChose.bind();
-  this.mediaComprehension = this.mediaComprehension.bind();
+  componentDidMount() {
+    const uri = `${redditAPI}${this.props.subreddit}.json`;
```
Comment (stcalica) [inline src/RedditPosts.js]: i like this!

File: src/RedditPosts.js [hunk 2]
```diff
@@ -1,153 +1,126 @@
-import React, { Component } from 'react';
-import RedditPostIt from './RedditPostIt';
-import RedditListing from './RedditListing';
-import RedditTile from './RedditTile';
+import React, { Component } from "react";
```
Comment (stcalica) [inline src/RedditPosts.js]: i always get confused on when to use single quotes and double quotes. 
in javascript and python!

File: src/RedditPosts.js [hunk 3]
```diff
@@ -1,153 +1,126 @@
...
-
-        let postData = {
-          key: index,
-          title:post.data.title,
-          media: this.mediaComprehension(post),
-          link: post.data.url
-        };
-        apiPosts.push(postData);
-      });
+  handlePosts(posts) {
```
Comment (stcalica) [inline src/RedditPosts.js]: much cleaner!

File: src/RedditPosts.js [hunk 4]
```diff
@@ -1,153 +1,126 @@
...
+        title: post.data.title,
+        media: this.getMediaFromPost(post),
+        link: post.data.url
+      };
+    });
+
+    this.setState({ redditPosts: apiPosts });
+  }
+
+  getMediaFromPost(post) {
```
Comment (stcalica) [inline src/RedditPosts.js]: yeah this is better than my extremely sloppy function!

**Review by stcalica — COMMENTED** [review]
Some questions: 

1) What advantages for moving `default export` in the beginning of the function and not at the end? Is the end for Node module's styling and in the beginning for React styling? 

2) When do you believe it is appropriate to use double quotes over single quotes? Is it the same conventions in Python? 

3)  I didn't know I could just make a ListComponent variable and store each of my components like that. I assumed I'd have to do inheritance down the road. this is much cleaner. Is this Pure Functions everyone was talking about? Could you explain this please? Like the return instead of render() in each of my display components(tile,listing,etc)? 


4)  Where do you think I can improve? I've been programming for years and my skills need to shape up. I can build things, but not as cleanly as I wished. Any recommendations for readings, practice, etc?

File: src/RedditPosts.js (hunk 2 above)
Comment (christianscott) [reply src/RedditPosts.js]: Apologies for the change from single to double - it 100% doesn't matter in JS! My formatter (prettier) uses double quotes by default. Use whatever tickles your fancy

**Comment by christianscott** [comment]
> What advantages for moving default export in the beginning of the function and not at the end? Is the end for Node module's styling and in the beginning for React styling?

This is quite subjective, but I think that it makes it much more obvious that a function/class is being exported. "export default" screams "pay attention to me". Having said that, it's not very important. If you prefer exports at the bottom of the file you should stick with that.

> When do you believe it is appropriate to use double quotes over single quotes? Is it the same conventions in Python?

As I said in the comment above this was a mistake - sorry! Singles are totally fine. The only important thing is consistency.

> I didn't know I could just make a ListComponent variable and store each of my components like that. I assumed I'd have to do inheritance down the road. this is much cleaner.

As with everything in Javascript, components are just data. You could do the same with a function or a class (which is also just a function).

Also, re. inheritance, you never need to do this in React!! [Read this article for some better patterns](https://reactjs.org/docs/composition-vs-inheritance.html).

> Is this Pure Functions everyone was talking about? Could you explain this please? Like the return instead of render() in each of my display components(tile,listing,etc)?

if you're talking about this:

```
export default function Name(props) {
	return <h1>{props.name}</h1>;
}
```

These are called "stateless functional components" (SFCs). They have the advantage of rendering in exactly the same way if they're given the same props. Are you familiar with the concept of a "pure function"? SFCs are reminiscent of this idea. It makes reasoning about what they're going to display a bit easier.

The above SFC effectively equivalent to this (bar some differences with how react handles them internally - not important):

```
export default class Name extends React.Component {
	render() {
		return <h1>{this.props.name}</h1>;
	}
}
```

The only difference is that props are on the instance rather than being passed as an argument.

> Where do you think I can improve? I've been programming for years and my skills need to shape up. I can build things, but not as cleanly as I wished. Any recommendations for readings, practice, etc?

I'm no expert, but I would say build things that are way out of your league, fail miserably, read other people's code. I think the key is to be perpetually confused 😆. 

Regarding specific materials, Clean Code by Robert C. Martin was fantastic. egghead.io also has some really great content.

**Comment by stcalica** [comment]
Your awesome thanks for your amazing advice as well! 

The sad thing is I'm an engineer full time and I'm so sad at how I've either gotten worse or never seemed to get better at programming. Guess I'll have to challenge myself even more! 

I'm happy for your contributions and will merge them now that I've tested them! 

I have a few more components I am writing out so I'll have a chance to practice what you've shown me here! Thank you.

**Pull request merged**